docs/adr/0006-consensus-governance.md

```diff
@@ -0,0 +1,455 @@
+  than `quorum`, the proposal is rejected.
+
+- In cae the percentage of `VoteYes` votes relative to all valid votes is less
+  than `threshold`, the proposal is rejected.
+
+- Otherwise the proposal is passed.
+
+- The proposal's status is changed to either `StatePassed` or `StateRejected`
+  and the proposal is saved.
+
+- The proposal is removed from the list of active proposals.
+
+- The corresponding `ProposalFinalizedEvent` is emitted with the following
+  structure:
+
+  ```golang
+  type ProposalFinalizedEvent struct {
+      // ID is the unique identifier of a proposal.
+      ID uint64 `json:"id"`
+      // State is the new proposal state.
+      State ProposalState `json:"state"`
+  }
+  ```
+
+- In case the proposal has been passed, the deposit is transferred back to the
+  proposal submitter and a corresponding `staking.TransferEvent` is emitted,
+  indicating transfer from the _proposal deposit account_ to the submitter's
+  account.
+
+- In case the proposal has been rejected, the deposit is transferred to the
+  common pool and a corresponding `staking.TransferEvent` is emitted,
+  indicating transfer from the _proposal deposit account_ to the common pool
+  account.
+
+- In case the proposal has been passed, the proposal content is executed.
+
+### Proposal Content Execution
+
+After any proposal is executed the corresponding `ProposalExecutedEvent` is
+emitted with the following structure:
+
+```golang
+type ProposalExecutedEvent struct {
+    // ID is the unique identifier of a proposal.
+    ID uint64 `json:"id"`
+}
+```
+
+#### Upgrade Proposal
+
+When an upgrade proposal is executed, a new entry is added to the list of
+pending upgrades using `epoch` as `<upgrade-epoch>`.
+
+On each epoch transition (as part of `BeginBlock`) it is checked wheter a
+pending upgrade is scheduled for that epoch. In case it is and we are not
+running the new version, the consensus layer will panic. Otherwise, the pending
+upgrade proposal is removed.
+
+#### Cancel Upgrade Proposal
+
+When a cancel upgrade proposal is executed, the proposal identified by
+`proposal_id` is looked up and removed from the list of pending upgrades. In
+case the pending upgrade does not exist anymore, no action is performed.
+
+### Consensus Parameters
+
+This proposal introduces the following new consensus parameters in the
+governance module:
+
+- `min_proposal_deposit` (base units) specifies the number of base units that
+  are deposited when creating a new proposal.
+
+- `voting_period` (number of blocks) specifies the number of blocks after which
+  the voting for a proposal is closed and the votes are tallied.
+
+- `quorum` (uint8) specifies the minimum percentage of voting power that needs
+  to be cast on a proposal for the result to be valid.
+
+- `threshold` (uint8) specifies the minimum percentage of `VoteYes` votes in
+  order for a proposal to be accepted.
+
+- `upgrade_min_epochs_diff` (epochs) specifies the minimum number of epochs
```

Comment:
Should there be a similar parameter for minimum diff between current height & proposal expiry, so something like `min_proposal_duration`?

Reply:
You mean the `voting_period` which is defined slightly above? Or something different?

Reply:
So when you create a proposal you only submit the `ProposalContent`, you can't define your own closing time etc.

Reply:
Oh right, yeah didn't notice that. Makes sense :+1: